README.markdown

```diff
@@ -150,6 +150,11 @@ Supported options:
 
 (I.e. you want only interface declarations for your Protobuf types, then pass all three of `outputEncodeMethods`, `outputJsonMethods`, and `outputClientImpl` as `false`, i.e. `--ts_proto_opt=outputEncodeMethods=false,outputJsonMethods=false,outputClientImpl=false`.)
 
+The following options are very useful for NestJS and available if `outputEncodeMethods`, `outputJsonMethods`, and `outputClientImpl` are set as `false`
```

Comment:
Just curious, but why do these other options need to be false? AFAIU it should be fine if these methods lived on the `FooMessage` entity, i.e. `FooMessage.toJSON/fromJSON`, but you go ahead and add `metadata` to the `Service` interface?

Reply:
To be truthful I didn't know if they were applicable, but let me have a look at adding them.

Reply:
Hi, could you give me some more details. What I have been able to do is add this information into the concrete types. I need to change the code contents a bit. Let me show an example of the generated code.

```
export interface HeroService {

  FindOne(request: HeroById, metadata: Metadata): Observable<Hero>;

  FindTwo(request: HeroById, metadata: Metadata): Observable<Elses>;

}

export class HeroServiceClientImpl implements HeroService {

  private readonly rpc: Rpc;

  constructor(rpc: Rpc) {
    this.rpc = rpc;
  }

  FindOne(request: HeroById, metadata: Metadata, request: HeroById): Observable<Hero> {
    const data = HeroById.encode(request).finish();
    const promise = this.rpc.request("hero.HeroService", "FindOne", data);
    return promise.then(data => Hero.decode(new Reader(data)));
  }

  FindTwo(request: HeroById, metadata: Metadata, request: HeroById): Observable<Elses> {
    const data = HeroById.encode(request).finish();
    const promise = this.rpc.request("hero.HeroService", "FindTwo", data);
    return promise.then(data => Elses.decode(new Reader(data)));
  }

}
```

So as you can see the metadata is there and it also returns an Observable, I need to update the lines 

```
    return promise.then(data => Elses.decode(new Reader(data)));
```

to

```
    from(promise).pipe(
         map((item) => {
           return Hero.decode(new Reader(data))
        }
    )
```

So the above would take the promise and return an observable. I need to check the above for syntax errors :-)

I didn't really understand the toJson / fromJson 

This should not need to change right ?

Reply:
The above concrete types would not be usable by NestJs - as it has its own implementation client library so only interfaces are required BUT I think the above would be useful for somebody using the standard grpc client library and wanted to work directly with observables.

Are you aware of anything else that should change.

Reply:
Or maybe this isn't a good idea adding it to the concrete types, if a person needs to have an observable from the existing promises that all that is needed is for them to do 

```
   from (...some promise)...
```

But then of course you would have an interface that returns an observable and a concrete type that returns a Promise 

:-(

Reply:
I haven't pushed these changed at the moment, I think it needs more discussion. I mean with the NestJs it is really needed as we need an interface that can directly offer an Observable. The interface is NOT implemented but uses as part of the NestJs client library - without this sort of functionality then you have to produce these by hand.

With regards to the concrete types, these can be called using the "from" command which converts a promise to an observable.

```
 from(some promise, from the concrete type).pipe(....do something else)
```

I think we need to think if it is really beneficial to people - I am not sure it would be.

What do you think @stephenh  ?